<a href="https://colab.research.google.com/github/salmanmfa2/data-science-2026/blob/main/pertemuan10_salman_mfa_240401010356.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

langkah 1: Load dan explore data

In [ ]:
import pandas as pd

df = pd.read_csv('https://raw.githubusercontent.com/IBM/telco-customer-churn-on-icp4d/d5371f5d83a446ad5673cbcca3b814b926491f8a/data/Telco-Customer-Churn.csv')
print(df.shape)
print(df.dtypes)
display(df.head())

(7043, 21)
customerID           object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges         object
Churn                object
dtype: object


,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


langkah 2: Preprocessing

In [ ]:

from sklearn.model_selection import train_test_split

# TODO: Pisahkan X (fitur) dan y (target = Churn)
X = df.drop("Churn", axis=1)
y = df["Churn"]

# TODO: Encoding fitur kategorikal menggunakan pd.get_dummies
X = pd.get_dummies(X, drop_first=True)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)

print(X_train.shape)
print(X_test.shape)
print( y_train.shape)
print(y_test.shape)


(5634, 13601)
(1409, 13601)
(5634,)
(1409,)


Langkah 3 : Latih Model

In [ ]:
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(n_estimators=300,class_weight="balanced",random_state=42)
rf.fit(X_train, y_train)

RandomForestClassifier(class_weight='balanced', n_estimators=300,
                       random_state=42)

Langkah 4 : Evaluasi

In [ ]:
from sklearn.metrics import classification_report, roc_auc_score
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score
import matplotlib.pyplot as plt
import seaborn as sns

# TODO: hitung prediksi dan probabilitas
y_pred = rf.predict(X_test)
y_prob = rf.predict_proba(X_test)[:, 1]
# TODO: tampilkan classification_report dan ROC-AUC
print(">> Classification Report <<")
print(classification_report(y_test, y_pred))
print(f"ROC-AUC Score: {roc_auc_score(y_test, y_prob):.4f}\n")

>> Classification Report <<
              precision    recall  f1-score   support

          No       0.84      0.90      0.87      1035
         Yes       0.65      0.52      0.58       374

    accuracy                           0.80      1409
   macro avg       0.74      0.71      0.72      1409
weighted avg       0.79      0.80      0.79      1409

ROC-AUC Score: 0.8337



Langkah 5: Prediksi Probabilitas dan Simpulkan

In [ ]:
# TODO: Hitung probabilitas churn (predict_proba)
churn_prediction = rf.predict_proba(X_test)[:, 1]
print("\Testing Churn dari Data Uji (10 Sampel)):")
print(churn_prediction[:10])

<>:3: SyntaxWarning: invalid escape sequence '\T'
<>:3: SyntaxWarning: invalid escape sequence '\T'
/tmp/ipykernel_2665/2582413704.py:3: SyntaxWarning: invalid escape sequence '\T'
  print("\Testing Churn dari Data Uji (10 Sampel)):")


\Testing Churn dari Data Uji (10 Sampel)):
[0.00333333 0.63666667 0.07333333 0.37666667 0.01666667 0.56666667
 0.46333333 0.03       0.00333333 0.53      ]


Model `RandomForestClassifier` dapat digunakan untuk memprediksi *churn* pelanggan Telco, dengan tahapan mulai dari memuat dan pra-pemrosesan data (termasuk *encoding* fitur dan penanganan nilai `TotalCharges`), pelatihan model dengan `class_weight='balanced'`, hingga evaluasi yang menghasilkan akurasi sekitar 80% dan skor ROC-AUC 0.8337, serta prediksi probabilitas *churn*. Model ini menunjukkan kinerja yang menjanjikan dalam mengidentifikasi pelanggan yang cenderung *churn*.